python_ta/checkers/bad_loop_indexing_checker.py

```diff
@@ -0,0 +1,109 @@
+"""checker for bad indexing in a loop.
+"""
+
+import astroid
+from pylint.interfaces import IAstroidChecker
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import check_messages
+from typing import List, Optional
+from astroid.node_classes import NodeNG
+
+
+class BadLoopIndexingChecker(BaseChecker):
+    __implements__ = IAstroidChecker
+    # name is the same as file name but without _checker part
+    name = 'bad_loop_indexing'
+    # use dashes for connecting words in message symbol
+    msgs = {'E9994': (f'Bad use of loop variable %s in for loop',
+                      'bad-loop-indexing',
+                      'Used when you have an loop variable in a for loop '
+                      'where its only usage is to index the iterable'),
+            }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    # pass in message symbol as a parameter of check_messages
+    @check_messages("bad-loop-indexing")
+    def visit_for(self, node: astroid.For) -> None:
+        if self._is_bad_loop_indexing(node):
+            args = f"'{node.target.name}'"
+            self.add_message('bad-loop-indexing', node=node.target, args=args)
+
+    def _is_bad_loop_indexing(self, node: astroid.For) -> bool:
+        """Return whether the usage of the indexing variable in the for loop is ONLY to index the iterable.
+        True if bad, False if not bad.
+        """
+        # Store the iteration variable.
+        index = node.target.name
+
+        # Check if the iterable of the for loop is of the form "range(len(lst))".
+        iterable = _iterable_if_range(node.iter)
+        if iterable is None:
+            return False
+
+        # AugAssign is special (i += 2 is actually i = i + 2).
+        # Check if any AssignName nodes in an AugAssign node refers to the iteration variable.
+        if any(index_node.lookup(index)[1][0] == node.target for index_node in _augassign_index_nodes(node, index)):
```

Comment:
The code is certainly improved from last time. However, the central algorithm can still be simplified, basically by merging the checks for `AugAssign` and `Name` nodes. Restructure your code as follows:

1. Obtain a list of *all* `AssignName` and `Name` nodes that have the same name as the loop target, and whose `.lookup` method call returns that target. Note that `nodes_of_class` can take in multiple node classes.
2. Check whether *all* of the nodes from the previous step are "redundant". This can be done using just a single helper function that takes a `AssignName` or `Name` node and determines whether it's "redundant"; also, this flips the boolean output from what you had before, so be careful here.
3. Return the correct thing. Remember to check for whether there's at least one usage of the loop target at all.

Note that this strategy will reduce the overall number of helper functions.